### PR TITLE
SHS-NG M4.2: Port executors page to new backend.

### PR DESCRIPTION
**Pull request by vanzin**
The executors page is built on top of the REST API, so the page itself
was easy to hook up to the new code.

Some other pages depend on the `ExecutorListener` class that is being
removed, though, so they needed to be modified to use data from the
new store. Fortunately, all they seemed to need is the map of executor
logs, so that was somewhat easy too.

The executor timeline graph required adding some properties to the
ExecutorSummary API type. Instead of following the previous code,
which stored all the listener events in memory, the timeline is
now created based on the data available from the API.

I had to change some of the test golden files because the old code would
return executors in "random" order (since it used a mutable Map instead
of something that returns a sorted list), and the new code returns executors
in id order.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -351,20 +355,26 @@ private[spark] class AppStatusListener(
...
+          exec.totalShuffleRead += readMetrics.localBytesRead + readMetrics.remoteBytesRead
+          exec.totalShuffleWrite += event.taskMetrics.shuffleWriteMetrics.bytesWritten
+        }
 
-      exec.activeTasks -= 1
-      exec.completedTasks += completedDelta
-      exec.failedTasks += failedDelta
-      exec.totalDuration += event.taskInfo.duration
-      liveUpdate(exec)
+        exec.activeTasks -= 1
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: shouldn't active tasks be updated even if `event.reason == Resubmitted`?

looks like that is what happens in the current ExecutorsTab

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 2]
```diff
@@ -123,55 +124,53 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
...
+           |{
+           |  'className': 'executor added',
+           |  'group': 'executors',
+           |  'start': new Date(${e.addTime.getTime()}),
+           |  'content': '<div class="executor-event-content"' +
+           |    'data-toggle="tooltip" data-placement="bottom"' +
+           |    'data-title="Executor ${e.id}<br>' +
+           |    'Added at ${UIUtils.formatDate(new Date(e.addTime.getTime()))}"' +
+           |    'data-html="true">Executor ${e.id} added</div>'
+           |}
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: what do you think about adding a helper method for these json snippets?  obviously not the fault of your change, but I find these so hard to read, eg. figuring out where `data-title` ends (worse in some of the other cases esp in StagePage).  something like

```scala
UIUtils.timelineJsonEvent(
  className="executor added",
  group="executors",
  start = e.addTime.getTime(),
  end = None,
  contentClass = "executor-event-content",
  dataTitle = s"Executor ${e.id}<br> Added at ${UIUtils.formatDate(new Date(e.addTime.getTime()))}",
  contentBody = s"Executor ${e.id} added")
```

**Review by squito — COMMENTED** [review]
just one small bug I think

again it would be good to mention where this is tested -- in this case I think existing tests actually cover this.

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 1 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Is it? This is the code right below the copy & pasted comment on top of this block:

```
      if (taskEnd.reason == Resubmitted) {
        return
      }
```

That being said, if a resubmitted task generates a `onTaskStart` call, then the accounting here will be off (and will also be off in the current code in master, as far as I can tell).

Comment (squito) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: oh you're right, sorry I missed the `return` -- its the same in master.

But yeah, master looks wrong too.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: I'm not against it but I'd rather do that separately.

**Pull request closed without merging**